sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala

```diff
@@ -220,7 +220,54 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
           if (attrMap.isEmpty) {
             planWithNewChildren
           } else {
-            planWithNewChildren.rewriteAttrs(attrMap)
+            def rewriteAttrsMatchWithSubPlan(
+                attrs: Seq[Attribute],
+                attrMap: AttributeMap[Attribute],
+                planOutput: Seq[Attribute]): Seq[Attribute] = {
+              val canRewriteAttrs = attrMap.filter(a => planOutput.contains(a._2))
+              attrs.map(attr => canRewriteAttrs.getOrElse(attr, attr))
+            }
+
+            def rewriteOrderMatchWithSubPlan(
+                attrs: Seq[SortOrder],
+                attrMap: AttributeMap[Attribute],
+                planOutput: Seq[Attribute]): Seq[SortOrder] = {
+              val canRewriteAttrs = attrMap.filter(a => planOutput.contains(a._2))
+              attrs.map(attr => {
+                attr.transformWithPruning(_.containsPattern(ATTRIBUTE_REFERENCE)) {
+                  case a: AttributeReference =>
+                    canRewriteAttrs.getOrElse(a, a)
+                }
+                attr
+              })
+            }
+
+            planWithNewChildren match {
+              case c: CoGroup =>
+                // SPARK-43781: CoGroup is a special case, as it has different output attributes
```

Comment:
What do you mean by "it has different output attributes"?

There are similar cases where it creates new output attributes (not from the child):  `MapPartitionsInR`, `MapPartitionsInRWithArrow`, `MapElements`, `MapGroups`, `FlatMapGroupsWithState`, `FlatMapGroupsInR`, `FlatMapGroupsInR`, `FlatMapGroupsInRWithArrow`, `FlatMapGroupsInPandas`,  `MapInPandas`, `PythonMapInArrow`,  and `FlatMapGroupsInPandasWithState`.

There's another CoGroup operation in Python (`cogroup.apply()` that is `FlatMapCoGroupsInPandas`). Would be great to double check.

Reply:
Yup, that's fine to do them as a followup.